src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -22,168 +23,314 @@ using namespace opossum::expression_functional;  // NOLINT
+    for (const auto& output : node->outputs()) {
+      const auto& required_expressions = required_expressions_by_node[output];
+      if (std::find_if(required_expressions.begin(), required_expressions.end(), [&expression_iter](const auto& other) {
+            return **expression_iter == *other;
+          }) != required_expressions.end()) {
+        new_node_expressions.emplace_back(*expression_iter);
+        break;
       }
     }
+  }
 
-    if (projection_node->node_expressions.size() == referenced_projection_expressions.size()) {
-      // No columns to prune
-      continue;
-    }
+  projection_node->node_expressions = new_node_expressions;
+}
+
+}  // namespace
+
+void ColumnPruningRule::apply_to(const std::shared_ptr<AbstractLQPNode>& lqp) const {
+  // For each node, required_expressions_by_node will hold the expressions either needed by this node or by one of its
+  // successors (i.e., nodes to which this node is an input). After collecting this information, we walk through all
+  // identified nodes and perform the pruning.
+  std::unordered_map<std::shared_ptr<AbstractLQPNode>, ExpressionUnorderedSet> required_expressions_by_node;
+
+  // Add top-level columns that need to be included as they are the actual output
+  required_expressions_by_node[lqp].insert(lqp->column_expressions().begin(), lqp->column_expressions().end());
+
+  // Recursively walk through the LQP. We cannot use visit_lqp as we explicitly need to take each path through the LQP.
+  // The right side of a diamond might require additional columns - if we only visited each node once, we might miss
+  // those. However, we track how many of a node's outputs we have already visited and recurse only once we have seen
+  // all of them. That way, the performance should be similar to that of visit_lqp.
+  std::unordered_map<std::shared_ptr<AbstractLQPNode>, size_t> outputs_visited_by_node;
+  recursive_gather_required_expressions(lqp, required_expressions_by_node, outputs_visited_by_node);
+
+  // Now, go through the LQP and perform all prunings. This time, it is sufficient to look at each node once.
+  for (const auto& [node, required_expressions] : required_expressions_by_node) {
+    switch (node->type) {
+      case LQPNodeType::StoredTable: {
+        // Prune all unused columns from a StoredTableNode
+        auto pruned_column_ids = std::vector<ColumnID>{};
+        for (const auto& expression : node->column_expressions()) {
+          if (required_expressions.find(expression) != required_expressions.end()) {
+            continue;
+          }
+
+          const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+          pruned_column_ids.emplace_back(column_expression->column_reference.original_column_id());
+        }
+
+        auto stored_table_node = std::dynamic_pointer_cast<StoredTableNode>(node);
+
+        if (pruned_column_ids.size() == stored_table_node->column_expressions().size()) {
```

Comment:
Can you give me a good example when this can happen? Looks rather random to leave the last row in the StoredTable node.

Reply:
> Can you give me a good example when this can happen?

`SELECT 1 FROM table` does not need any particular column, but it needs at least one column so that it knows how many 1s to produce. Will add this as a comment, please don't resolve.

> Looks rather random to leave the last row in the StoredTable node.

I can't follow. What do you mean by "last row"?

Reply:
Last row should be "last column". But I misread your code tbh. What are you actually doing here? In case we could potentially prune all columns (which might result in a failure), we remove just the last column?

And what if `stored_table_node->column_expressions().size()` is 1 and the all columns are prune. Then you would resize to 0, right? Which will fail?

Reply:
* Add comment from above
* s/Cannot prune all columns/Cannot prune all columns - remove one column from pruning list so that at least one column remains/